elastic/apm-server
# Direct agent configuration

## Motivation/summary

Add support for direct agent configuration in apm-server

## Checklist


## How to test these changes

Build and run `apm-server` with `agent_configs` defined. `cURL` the server to verify the response makes sense for the incoming request.

`apm-server.yml`
```yaml
  default_service_environment: production
  agent_config:
  - service.name: ten_percent
    config:
      capture_body: off
      capture_body: true
      log_level: info
      recording: true
  - service.environment: production
    config:
      transaction_sample_rate: 0.5
  - service.environment: dev
    config:
      transaction_sample_rate: 0.1

```
verify the default_service_environment is respected
```
# curl 'http://localhost:8200/config/v1/agents?service.name=ten_percent'
{
  "capture_body": "true",
  "log_level": "info",
  "recording": "true",
  "transaction_sample_rate": "0.5"
}
```
verify alternative environments can be requested
```
# curl 'http://localhost:8200/config/v1/agents?service.name=ten_percent&service.environment=dev'
{
  "capture_body": "true",
  "log_level": "info",
  "recording": "true",
  "transaction_sample_rate": "0.1"
}
```
requests with a missing or unrecognized `service.name` return an error
```
# curl 'http://localhost:8200/config/v1/agents?service.name=dne'
{
  "error": "no config found for service.name=dne"
}
```
## Related issues

closes https://github.com/elastic/apm-server/issues/5018
